SQL ICE: value borrowed after move

```
-- Forward declaration of spreadsheet view
create recursive view spreadsheet_view (
    id bigint not null,
    raw_value varchar(64) not null,
    background integer not null,
    computed_value varchar(64) not null
);

-- Forward declaration of mentions_with_values view
create recursive view mentions_with_values (
    id BIGINT NOT NULL,
    raw_value VARCHAR(64) NOT NULL,
    background INTEGER NOT NULL
);

-- Raw spreadsheet cell data coming from backend/user, updates
-- are inserted as new entries with newer timestamps
create table spreadsheet_data (
    id bigint not null,
    ip varchar(45) not null,
    ts timestamp not null,
    raw_value varchar(64) not null,
    background integer not null
) with (
      'materialized' = 'true',
      'connectors' = '[{
        "transport": {
            "name": "datagen",
            "config": {
                "workers": 1,
                "plan": [{
                    "rate": 1,
                    "limit": 2,
                    "fields": {
                        "id": { "range": [0, 1000] },
                        "ip": { "strategy": "uniform", "range": [0, 5] },
                        "raw_value": { "values": ["=B0", "0"] },
                        "background": { "strategy": "uniform", "range": [0, 1000] }
                    }
                }]
            }
        }
    }]'
      );

-- Get the latest cell value for the spreadsheet.
-- (By finding the one with the highest `ts` for a given `id`)
create view latest_cells as with
    max_ts_per_cell as (
        select
            id,
            max(ts) as max_ts
        from
            spreadsheet_data
        group by
            id
    )
select
    s.id,
    s.raw_value,
    s.background
from
    spreadsheet_data s
        join max_ts_per_cell mt on s.id = mt.id and s.ts = mt.max_ts;

-- List all mentioned ids per latest cell
create view latest_cells_with_mentions as
select
    s.id,
    s.raw_value,
    s.background
from
    latest_cells s;

-- Like latest_cells_with_mentions, but enrich it with values of mentioned cells
create view mentions_with_values as
select
    m.id,
    m.raw_value,
    m.background
from
    latest_cells_with_mentions m;

-- We aggregate mentioned values and ids back into arrays
create view mentions_aggregated as
select
    id,
    raw_value,
    background
from
    mentions_with_values
group by
    id,
    raw_value,
    background;

-- Calculate the final spreadsheet by executing the UDF for the formula
create materialized view spreadsheet_view as
select
    id,
    raw_value,
    background,
    raw_value AS computed_value
from
    mentions_aggregated;
```

leads to

```

Rust error: the Rust code generated based on the SQL failed to compile, which should not happen. Please file a bug report with the example SQL that triggers it at: https://github.com/feldera/feldera/issues

The compilation task produced the following output:

stdout:

stderr:
   Compiling project01934121-eef8-76f2-96fa-ad5fd53604fa v0.1.0 (/media/gz/8679599d-3ee6-43e8-bdd6-1ad7a3714008/feldera/cargo_workspace/project01934121-eef8-76f2-96fa-ad5fd53604fa)
error[E0382]: borrow of moved value: `s5`
   --> project01934121-eef8-76f2-96fa-ad5fd53604fa/src/main.rs:280:54
    |
220 |         let s5: Stream<_, WSet<Tup3<i64, String, i32>>> = s3.join(&s4, move |t_17: &Tup2<i64, Timestamp>, t_18: &Tup2<String, i32>, t_19:...
    |             -- move occurs because `s5` has type `dbsp::Stream<ChildCircuit<()>, TypedBatch<dbsp::utils::Tup3<i64, std::string::String, i32>, (), i64, DynFallbackWSet<dyn dbsp::dynamic::Data, dyn dbsp::dynamic::weight::WeightTyped<Type = i64>>>>`, which does not implement the `Copy` trait
...
227 |         let s6 = s5;
    |                  -- value moved here
...
280 |         catalog.register_output_zset::<_, struct_17>(s5.clone(), r#"{"name":"latest_cells","case_sensitive":false,"fields":[{"name":"id",...
    |                                                      ^^ value borrowed here after move
    |
help: consider cloning the value if the performance cost is acceptable
    |
227 |         let s6 = s5.clone();
    |                    ++++++++

For more information about this error, try `rustc --explain E0382`.
error: could not compile `project01934121-eef8-76f2-96fa-ad5fd53604fa` (bin "project01934121-eef8-76f2-96fa-ad5fd53604fa") due to 1 previous error


Exit code: exit status: 101
```

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/CircuitToString.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/CircuitToString.java
@@ -25,6 +25,15 @@ public class CircuitToString {
         for (var child: operator.getAllOperators()) {
             toString(child);
         }
+        int index = 0;
+        for (var o: operator.outputs) {
+            this.builder.append(operator.id)
+                    .append(":")
+                    .append(index++)
+                    .append(" = ");
+            port(o);
+            this.builder.newline();
+        }
         builder.decrease().append("}").newline();
     }
 

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/operator/DBSPNestedOperator.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/operator/DBSPNestedOperator.java
@@ -12,6 +12,7 @@ import org.dbsp.util.IIndentStream;
 import org.dbsp.util.Linq;
 import org.dbsp.util.Utilities;
 
+import javax.annotation.Nullable;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -24,11 +25,14 @@ public class DBSPNestedOperator extends DBSPOperator implements ICircuit {
     final List<DBSPOperator> allOperators;
     final Set<DBSPOperator> operators;
     final Map<String, DBSPViewOperator> viewByName;
-    public final List<DBSPViewDeclarationOperator> viewDeclarations;
+    /** Indexed by original view name */
+    public final Map<String, DBSPViewDeclarationOperator> declarationByName;
     final List<DBSPDeltaOperator> deltaInputs;
     /** For each output port of this, the actual port of an operator inside,
      * which produces the result. */
     public final List<OutputPort> outputs;
+    /** Outputs always correspond to recursive views.  Names of these views in order */
+    public final List<String> outputViews;
 
     public DBSPNestedOperator(CalciteObject node) {
         super(node);
@@ -37,7 +41,8 @@ public class DBSPNestedOperator extends DBSPOperator implements ICircuit {
         this.deltaInputs = new ArrayList<>();
         this.outputs = new ArrayList<>();
         this.operators = new HashSet<>();
-        this.viewDeclarations = new ArrayList<>();
+        this.outputViews = new ArrayList<>();
+        this.declarationByName = new HashMap<>();
     }
 
     public boolean contains(DBSPOperator operator) {
@@ -62,11 +67,13 @@ public class DBSPNestedOperator extends DBSPOperator implements ICircuit {
             this.deltaInputs.add(delta);
             this.addInput(delta.input());
         } else if (operator.is(DBSPViewDeclarationOperator.class)) {
-            this.viewDeclarations.add(operator.to(DBSPViewDeclarationOperator.class));
+            var decl = operator.to(DBSPViewDeclarationOperator.class);
+            Utilities.putNew(this.declarationByName, decl.originalViewName(), decl);
         }
     }
 
-    public OutputPort addOutput(OutputPort port) {
+    public OutputPort addOutput(String view, OutputPort port) {
+        this.outputViews.add(view);
         this.outputs.add(port);
         assert this.operators.contains(port.node());
         return new OutputPort(this, this.outputs.size() - 1);
@@ -77,9 +84,10 @@ public class DBSPNestedOperator extends DBSPOperator implements ICircuit {
         throw new InternalCompilerError("Adding declaration to CircuitOperator");
     }
 
+    @Nullable
     @Override
     public DBSPViewOperator getView(String name) {
-        return Utilities.getExists(this.viewByName, name);
+        return this.viewByName.get(name);
     }
 
     @Override

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/DBSPCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/DBSPCompiler.java
@@ -527,7 +527,7 @@ public class DBSPCompiler implements IWritesLogs, ICompilerComponent, IErrorRepo
 
             this.frontend.endCompilation(this.compiler());
             this.circuit = this.midend.getFinalCircuit().seal("parsed");
-            if (this.getDebugLevel() > 2) {
+            if (this.getDebugLevel() > 1) {
                 ToDot.dump(this, "initial.png", this.getDebugLevel(), "png", this.circuit);
             }
 

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustVisitor.java
@@ -484,33 +484,40 @@ public class ToRustVisitor extends CircuitVisitor {
     }
 
     @Override
-    public VisitDecision preorder(DBSPNestedOperator circuit) {
-        boolean recursive = circuit.hasAnnotation(a -> a.is(Recursive.class));
+    public VisitDecision preorder(DBSPNestedOperator operator) {
+        boolean recursive = operator.hasAnnotation(a -> a.is(Recursive.class));
         if (!recursive)
             throw new InternalCompilerError("NestedOperator not recursive");
 
         this.builder.append("let (");
-        for (int i = 0; i < circuit.outputCount(); i++) {
-            this.builder.append(circuit.getOutputName(i)).append(", ");
+        for (int i = 0; i < operator.outputCount(); i++) {
+            this.builder.append(operator.getOutputName(i)).append(", ");
         }
         this.builder.append(") = ")
                 .append("circuit.recursive(|child, (");
-        for (DBSPViewDeclarationOperator decl: circuit.viewDeclarations) {
-            this.builder.append(decl.getOutputName()).append(", ");
+        for (int i = 0; i < operator.outputCount(); i++) {
+            String view = operator.outputViews.get(i);
+            DBSPViewDeclarationOperator decl = operator.declarationByName.get(view);
+            if (decl != null) {
+                this.builder.append(decl.getOutputName()).append(", ");
+            } else {
+                // view is not really recursive
+                this.builder.append("_").append(", ");
+            }
         }
         this.builder.append("): (");
-        for (int i = 0; i < circuit.outputCount(); i++) {
-            circuit.streamType(i).accept(this.innerVisitor);
+        for (int i = 0; i < operator.outputCount(); i++) {
+            operator.streamType(i).accept(this.innerVisitor);
             this.builder.append(", ");
         }
         this.builder.append(")| {").increase().newline();
-        for (IDBSPNode node : circuit.getAllOperators())
+        for (IDBSPNode node : operator.getAllOperators())
             this.processNode(node);
 
         this.builder.append("Ok((");
-        for (int i = 0; i < circuit.outputCount(); i++) {
+        for (int i = 0; i < operator.outputCount(); i++) {
             this.builder
-                    .append(circuit.outputs.get(i).getOutputName())
+                    .append(operator.outputs.get(i).getOutputName())
                     .append(", ");
         }
         this.builder.append("))").newline()

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
@@ -126,7 +126,6 @@ import java.nio.charset.Charset;
 import java.util.List;
 import java.util.Objects;
 
-import static org.apache.calcite.util.BuiltInMethod.SPLIT_PART;
 import static org.dbsp.sqlCompiler.ir.type.DBSPTypeCode.NULL;
 import static org.dbsp.sqlCompiler.ir.type.DBSPTypeCode.USER;
 
@@ -227,7 +226,7 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
                 String str = literal.getValueAs(String.class);
                 RelDataType litType = literal.getType();
                 Charset charset = litType.getCharset();
-                return new DBSPStringLiteral(Objects.requireNonNull(str), Objects.requireNonNull(charset));
+                return new DBSPStringLiteral(Objects.requireNonNull(str), Objects.requireNonNull(charset), type);
             }
             else if (type.is(DBSPTypeBool.class))
                 return new DBSPBoolLiteral(Objects.requireNonNull(literal.getValueAs(Boolean.class)));

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/CircuitCloneVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/CircuitCloneVisitor.java
@@ -417,7 +417,7 @@ public class CircuitCloneVisitor extends CircuitVisitor implements IWritesLogs {
             OutputPort originalOutput = operator.outputs.get(i);
             OutputPort newPort = this.mapped(originalOutput);
             if (result != operator) {
-                result.addOutput(newPort);
+                result.addOutput(operator.outputViews.get(i), newPort);
             }
             this.map(new OutputPort(operator, i), new OutputPort(result, i), false);
         }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/CircuitOptimizer.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/CircuitOptimizer.java
@@ -97,7 +97,7 @@ public record CircuitOptimizer(DBSPCompiler compiler) implements ICompilerCompon
         passes.add(new ExpandHop(reporter));
         passes.add(new RemoveDeindexOperators(reporter));
         passes.add(new OptimizeWithGraph(reporter, g -> new RemoveNoops(reporter, g)));
-        passes.add(new RemoveViewOperators(reporter));
+        passes.add(new RemoveViewOperators(reporter, false));
         passes.add(new Repeat(reporter, new ExpandCasts(reporter).circuitRewriter()));
         if (options.languageOptions.optimizationLevel >= 2) {
             passes.add(new OptimizeWithGraph(reporter, g -> new FilterMapVisitor(reporter, g)));
@@ -118,8 +118,9 @@ public record CircuitOptimizer(DBSPCompiler compiler) implements ICompilerCompon
         // Lowering may surface additional casts that need to be expanded
         passes.add(new Repeat(reporter, new ExpandCasts(reporter).circuitRewriter()));
         // Beta reduction after implementing aggregates.
-        passes.add(new BetaReduction(compiler).getCircuitVisitor());
-        passes.add(new CompactNames(compiler));
+        passes.add(new BetaReduction(reporter).getCircuitVisitor());
+        passes.add(new RemoveViewOperators(reporter, true));
+        passes.add(new CompactNames(reporter));
         return new Passes("optimize", reporter, passes);
     }
 

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/OptimizeIncrementalVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/OptimizeIncrementalVisitor.java
@@ -257,7 +257,7 @@ public class OptimizeIncrementalVisitor extends CircuitCloneVisitor {
             OutputPort originalOutput = operator.outputs.get(i);
             OutputPort newPort = this.mapped(originalOutput);
             if (result != operator) {
-                result.addOutput(newPort);
+                result.addOutput(operator.outputViews.get(i), newPort);
             }
             // The integrator receives the input from 'result', not from 'newPort'
             DBSPIntegrateOperator integral = new DBSPIntegrateOperator(

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/RecursiveComponents.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/RecursiveComponents.java
@@ -57,6 +57,7 @@ public class RecursiveComponents extends Passes {
         this.add(graph2);
         this.add(new BuildNestedOperators(reporter, graph2.getGraphs()));
         this.add(new ValidateRecursiveOperators(reporter));
+        // this.add(new ShowCircuit(reporter));
     }
 
     /** Check that all operators in recursive components are supported */
@@ -209,8 +210,15 @@ public class RecursiveComponents extends Passes {
                         } else if (source.is(DBSPViewOperator.class)) {
                             var view = source.to(DBSPViewOperator.class);
                             if (view.metadata.recursive && sourceScc == sccId) {
-                                var decl = Utilities.getExists(declByName, view.viewName);
-                                newPort = decl.outputPort();
+                                var decl = declByName.get(view.viewName);
+                                if (decl != null) {
+                                    // decl can be null if the view is declared recursive, but in fact it is not
+                                    // used recursively in an SCC.  The view may still need to be declared recursive
+                                    // because the user may want to e.g., materialize it.
+                                    newPort = decl.outputPort();
+                                } else {
+                                    newPort = this.mapped(view.outputPort());
+                                }
                             }
                         }
                         newSources.add(newPort);
@@ -232,10 +240,13 @@ public class RecursiveComponents extends Passes {
                                     view.getNode());
                         }
                         Utilities.putNew(viewByName, view.viewName, view);
-                    }
-                    if (result.is(DBSPViewDeclarationOperator.class)) {
+                    } else if (result.is(DBSPViewDeclarationOperator.class)) {
                         DBSPViewDeclarationOperator view = result.to(DBSPViewDeclarationOperator.class);
                         Utilities.putNew(declByName, view.originalViewName(), view);
+                        if (operators.size() == 1) {
+                            // If the declaration is in an SCC by itself, remove it from the graph.
+                            continue;
+                        }
                     }
                     result.setDerivedFrom(operator.id);
                     this.map(simple, result, true);
@@ -257,11 +268,14 @@ public class RecursiveComponents extends Passes {
         SCC<DBSPOperator> scc = null;
         final Map<Integer, DBSPNestedOperator> components;
         final Set<DBSPNestedOperator> toAdd;
+        /** Maps each view in an SCC to its output. */
+        final Map<String, OutputPort> viewPort;
 
         BuildNestedOperators(IErrorReporter reporter, CircuitGraphs graphs) {
             super(reporter, graphs, false);
             this.components = new HashMap<>();
             this.toAdd = new HashSet<>();
+            this.viewPort = new HashMap<>();
         }
 
         @Override
@@ -290,13 +304,13 @@ public class RecursiveComponents extends Passes {
                 }
             }
 
-            // Check if any inputs of the operator are in a different component
-            // If they are, insert a delta + integrator operator in front.
             List<OutputPort> sources = new ArrayList<>();
             for (OutputPort input : operator.inputs) {
                 OutputPort source = this.mapped(input);
                 int sourceComp = Utilities.getExists(this.scc.componentId, input.node());
                 if (sourceComp != myComponent) {
+                    // Check if any inputs of the operator are in a different component
+                    // If they are, insert a delta + integrator operator in front.
                     DBSPDeltaOperator delta = new DBSPDeltaOperator(operator.getNode(), source);
                     block.addOperator(delta);
                     DBSPIntegrateOperator integral = new DBSPIntegrateOperator(operator.getNode(), delta.outputPort());
@@ -307,6 +321,16 @@ public class RecursiveComponents extends Passes {
                     DBSPIntegrateOperator integral = new DBSPIntegrateOperator(operator.getNode(), source);
                     block.addOperator(integral);
                     sources.add(integral.outputPort());
+                } else if (input.node().is(DBSPViewOperator.class)) {
+                    // Same component and input is a view.  Can't use 'source', because
+                    // that is mapped to the differentiator in the check below if the view is recursive.
+                    DBSPViewOperator view = input.node().to(DBSPViewOperator.class);
+                    if (view.metadata.recursive) {
+                        OutputPort port = Utilities.getExists(this.viewPort, view.viewName);
+                        sources.add(port);
+                    } else {
+                        sources.add(source);
+                    }
                 } else {
                     sources.add(source);
                 }
@@ -323,7 +347,8 @@ public class RecursiveComponents extends Passes {
                 DBSPDifferentiateOperator diff = new DBSPDifferentiateOperator(operator.getNode(), view.outputPort());
                 block.addOperator(diff);
                 // Use as port the output of the NestedOperator, and not the source view
-                port = block.addOutput(diff.outputPort());
+                port = block.addOutput(view.viewName, diff.outputPort());
+                Utilities.putNew(this.viewPort, view.viewName, view.outputPort());
             }
             this.map(operator.outputPort(), port, false);
         }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/RemoveViewOperators.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/RemoveViewOperators.java
@@ -4,15 +4,21 @@ import org.dbsp.sqlCompiler.circuit.operator.DBSPViewOperator;
 import org.dbsp.sqlCompiler.circuit.OutputPort;
 import org.dbsp.sqlCompiler.compiler.IErrorReporter;
 
-/** Remove DBSPViewOperator operators that are not recursive. */
+/** Remove DBSPViewOperator operators. */
 public class RemoveViewOperators extends CircuitCloneVisitor {
-    public RemoveViewOperators(IErrorReporter reporter) {
+    /** If false remove only non-recursive views */
+    final boolean all;
+
+    public RemoveViewOperators(IErrorReporter reporter, boolean all) {
         super(reporter, false);
+        this.all = all;
     }
 
     @Override
     public void postorder(DBSPViewOperator operator) {
-        if (operator.metadata.recursive) {
+        if (!this.all && operator.metadata.recursive) {
+            // preserve the view.
+            // Views have to be preserved if we build the CircuitGraph.
             super.postorder(operator);
             return;
         }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/literal/DBSPStringLiteral.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/literal/DBSPStringLiteral.java
@@ -44,10 +44,6 @@ public final class DBSPStringLiteral extends DBSPLiteral {
     public final String value;
     public final Charset charset;
 
-    public DBSPStringLiteral(String value, Charset charset) {
-        this(value, charset, DBSPTypeString.varchar(false));
-    }
-
     public DBSPStringLiteral(String value) {
         this(value, StandardCharsets.UTF_8, DBSPTypeString.varchar(false));
     }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/recursive/RecursiveTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/recursive/RecursiveTests.java
@@ -1,6 +1,7 @@
 package org.dbsp.sqlCompiler.compiler.sql.recursive;
 
 import org.dbsp.sqlCompiler.circuit.operator.DBSPNestedOperator;
+import org.dbsp.sqlCompiler.circuit.operator.DBSPSinkOperator;
 import org.dbsp.sqlCompiler.compiler.StderrErrorReporter;
 import org.dbsp.sqlCompiler.compiler.sql.tools.BaseSQLTests;
 import org.dbsp.sqlCompiler.compiler.visitors.outer.CircuitVisitor;
@@ -9,6 +10,418 @@ import org.junit.Test;
 
 /** Tests with recursive queries */
 public class RecursiveTests extends BaseSQLTests {
+    @Test
+    public void issue2979() {
+        String sql = """
+                -- Given a cell value as a formula (e.g., =A0+B0), and a context with cell values
+                -- referenced in the formula, returns the computed value of the cell
+                create function cell_value(cell varchar(64), mentions_ids bigint array, mentions_values varchar(64) array) returns varchar(64);
+                
+                -- Given a cell value e.g., =A0+B0, returns an array of cell ids that were mentioned in the formula
+                create function mentions(cell varchar(64)) returns bigint array;
+                
+                -- Forward declaration of spreadsheet view
+                create recursive view spreadsheet_view (
+                    id bigint not null,
+                    raw_value varchar(64) not null,
+                    background integer not null,
+                    mentions_ids bigint array,
+                    mentions_values varchar(64) array,
+                    computed_value varchar(64)
+                );
+                
+                -- Forward declaration of mentions_with_values view
+                -- create recursive view mentions_with_values (
+                --     id BIGINT NOT NULL,
+                --     raw_value VARCHAR(64) NOT NULL,
+                --     background INTEGER NOT NULL,
+                --     mentioned_id BIGINT,
+                --     mentioned_value VARCHAR(64)
+                -- );
+                
+                -- Raw spreadsheet cell data coming from backend/user, updates
+                -- are inserted as new entries with newer timestamps
+                create table spreadsheet_data (
+                    id bigint not null,
+                    ip varchar(45) not null,
+                    ts timestamp not null,
+                    raw_value varchar(64) not null,
+                    background integer not null
+                ) with (
+                      'materialized' = 'true',
+                      'connectors' = '[{
+                        "transport": {
+                            "name": "datagen",
+                            "config": {
+                                "workers": 1,
+                                "plan": [{
+                                    "rate": 1,
+                                    "limit": 2,
+                                    "fields": {
+                                        "id": { "range": [0, 1000] },
+                                        "ip": { "strategy": "uniform", "range": [0, 5] },
+                                        "raw_value": { "values": ["=B0", "0"] },
+                                        "background": { "strategy": "uniform", "range": [0, 1000] }
+                                    }
+                                }]
+                            }
+                        }
+                    }]'
+                      );
+                
+                -- Get the latest cell value for the spreadsheet.
+                -- (By finding the one with the highest `ts` for a given `id`)
+                create view latest_cells as with
+                    max_ts_per_cell as (
+                        select
+                            id,
+                            max(ts) as max_ts
+                        from
+                            spreadsheet_data
+                        group by
+                            id
+                    )
+                select
+                    s.id,
+                    s.raw_value,
+                    s.background,
+                    -- The append with null is silly but crucial to ensure that the
+                    -- cross join in `latest_cells_with_mention` returns all cells
+                    -- not just those that reference another cell
+                    ARRAY_APPEND(mentions(s.raw_value), null) as mentioned_cell_ids
+                from
+                    spreadsheet_data s
+                        join max_ts_per_cell mt on s.id = mt.id and s.ts = mt.max_ts;
+                
+                -- List all mentioned ids per latest cell
+                create view latest_cells_with_mentions as
+                select
+                    s.id,
+                    s.raw_value,
+                    s.background,
+                    m.mentioned_id
+                from
+                    latest_cells s, unnest(s.mentioned_cell_ids) as m(mentioned_id);
+                
+                -- Like latest_cells_with_mentions, but enrich it with values of mentioned cells
+                create local view mentions_with_values as
+                select
+                    m.id,
+                    m.raw_value,
+                    m.background,
+                    m.mentioned_id,
+                    sc.computed_value as mentioned_value
+                from
+                    latest_cells_with_mentions m
+                left join
+                    spreadsheet_view sc on m.mentioned_id = sc.id;
+                
+                -- We aggregate mentioned values and ids back into arrays
+                create local view mentions_aggregated as
+                select
+                    id,
+                    raw_value,
+                    background,
+                    ARRAY_AGG(mentioned_id) as mentions_ids,
+                    ARRAY_AGG(mentioned_value) as mentions_values
+                from
+                    mentions_with_values
+                group by
+                    id,
+                    raw_value,
+                    background;
+                
+                -- Calculate the final spreadsheet by executing the UDF for the formula
+                create materialized view spreadsheet_view as
+                select
+                    id,
+                    raw_value,
+                    background,
+                    mentions_ids,
+                    mentions_values,
+                    cell_value(raw_value, mentions_ids, mentions_values) AS computed_value
+                from
+                    mentions_aggregated;
+                
+                -- Figure out which IPs currently reached their API limit
+                create materialized view api_limit_reached as
+                select
+                    ip,
+                    count(*) as request_count
+                from
+                    spreadsheet_data
+                where
+                    ts >= NOW() - INTERVAL 60 MINUTES
+                group by
+                    ip
+                having
+                    count(*) > 200;
+                
+                -- Compute statistics
+                create materialized view spreadsheet_statistics as
+                with filled_total as (
+                    select
+                        count(distinct id) as filled_total
+                    from
+                        spreadsheet_data
+                ),
+                filled_this_hour as (
+                    select
+                        count(*) as filled_this_hour
+                    from
+                        spreadsheet_data
+                    where
+                        ts >= NOW() - INTERVAL 1 HOUR
+                ),
+                filled_today as (
+                    select
+                        count(*) as filled_today
+                    from
+                        spreadsheet_data
+                    where
+                        ts >= NOW() - INTERVAL 1 DAY
+                ),
+                filled_this_week as (
+                    select
+                        count(*) as filled_this_week
+                    from
+                        spreadsheet_data
+                    where
+                        ts >= NOW() - INTERVAL 1 WEEK
+                ),
+                currently_active_users as (
+                    select
+                        count(distinct ip) as currently_active_users
+                    from
+                        spreadsheet_data
+                    where
+                        ts >= NOW() - INTERVAL 5 MINUTE
+                )
+                select
+                    (select filled_total from filled_total) as filled_total,
+                    (select filled_this_hour from filled_this_hour) as filled_this_hour,
+                    (select filled_today from filled_today) as filled_today,
+                    (select filled_this_week from filled_this_week) as filled_this_week,
+                    (select currently_active_users from currently_active_users) as currently_active_users;""";
+        var ccs = this.getCCS(sql);
+        CircuitVisitor visitor = new CircuitVisitor(new StderrErrorReporter()) {
+            @Override
+            public void postorder(DBSPSinkOperator operator) {
+                assert !operator.viewName.endsWith("-port");
+            }
+        };
+        visitor.apply(ccs.circuit);
+    }
+
+    @Test
+    public void issue2978() {
+        String sql = """
+                -- Forward declaration of spreadsheet view
+                create recursive view spreadsheet_view (
+                    id bigint not null,
+                    raw_value varchar(64) not null,
+                    background integer not null,
+                    computed_value varchar(64) not null
+                );
+                
+                -- Forward declaration of mentions_with_values view
+                create recursive view mentions_with_values (
+                    id BIGINT NOT NULL,
+                    raw_value VARCHAR(64) NOT NULL,
+                    background INTEGER NOT NULL
+                );
+                
+                -- Raw spreadsheet cell data coming from backend/user, updates
+                -- are inserted as new entries with newer timestamps
+                create table spreadsheet_data (
+                    id bigint not null,
+                    ip varchar(45) not null,
+                    ts timestamp not null,
+                    raw_value varchar(64) not null,
+                    background integer not null
+                );
+                
+                -- Get the latest cell value for the spreadsheet.
+                -- (By finding the one with the highest `ts` for a given `id`)
+                create view latest_cells as with
+                    max_ts_per_cell as (
+                        select
+                            id,
+                            max(ts) as max_ts
+                        from
+                            spreadsheet_data
+                        group by
+                            id
+                    )
+                select
+                    s.id,
+                    s.raw_value,
+                    s.background
+                from
+                    spreadsheet_data s
+                        join max_ts_per_cell mt on s.id = mt.id and s.ts = mt.max_ts;
+                
+                -- List all mentioned ids per latest cell
+                create view latest_cells_with_mentions as
+                select
+                    s.id,
+                    s.raw_value,
+                    s.background
+                from
+                    latest_cells s;
+                
+                -- Like latest_cells_with_mentions, but enrich it with values of mentioned cells
+                create view mentions_with_values as
+                select
+                    m.id,
+                    m.raw_value,
+                    m.background
+                from
+                    latest_cells_with_mentions m;
+                
+                -- We aggregate mentioned values and ids back into arrays
+                create view mentions_aggregated as
+                select
+                    id,
+                    raw_value,
+                    background
+                from
+                    mentions_with_values
+                group by
+                    id,
+                    raw_value,
+                    background;
+                
+                -- Calculate the final spreadsheet by executing the UDF for the formula
+                create materialized view spreadsheet_view as
+                select
+                    id,
+                    raw_value,
+                    background,
+                    raw_value AS computed_value
+                from
+                    mentions_aggregated;""";
+        var ccs = this.getCCS(sql);
+        this.addRustTestCase(ccs);
+    }
+
+    @Test
+    public void issue2980() {
+        String sql = """
+                -- Given a cell value as a formula (e.g., =A0+B0), and a context with cell values
+                -- referenced in the formula, returns the computed value of the cell
+                create function cell_value(cell varchar(64), mentions_ids bigint array, mentions_values varchar(64) array)
+                returns varchar(64) not null as CAST('' AS VARCHAR(64));
+                
+                -- Given a cell value e.g., =A0+B0, returns an array of cell ids that were mentioned in the formula
+                create function mentions(cell varchar(64)) returns bigint array not null as CAST(array(null) AS BIGINT ARRAY);
+                
+                -- Forward declaration of spreadsheet view
+                create recursive view spreadsheet_view (
+                    id bigint not null,
+                    raw_value varchar(64) not null,
+                    background integer not null,
+                    mentions_ids bigint array,
+                    mentions_values varchar(64) array,
+                    computed_value varchar(64) not null
+                );
+                
+                -- Forward declaration of mentions_with_values view
+                create recursive view mentions_with_values (
+                    id BIGINT NOT NULL,
+                    raw_value VARCHAR(64) NOT NULL,
+                    background INTEGER NOT NULL,
+                    mentioned_id BIGINT,
+                    mentioned_value VARCHAR(64)
+                );
+                
+                -- Raw spreadsheet cell data coming from backend/user, updates
+                -- are inserted as new entries with newer timestamps
+                create table spreadsheet_data (
+                    id bigint not null,
+                    ip varchar(45) not null,
+                    ts timestamp not null,
+                    raw_value varchar(64) not null,
+                    background integer not null
+                ) with ( 'materialized' = 'true' );
+                
+                -- Get the latest cell value for the spreadsheet.
+                -- (By finding the one with the highest `ts` for a given `id`)
+                create view latest_cells as with
+                    max_ts_per_cell as (
+                        select
+                            id,
+                            max(ts) as max_ts
+                        from
+                            spreadsheet_data
+                        group by
+                            id
+                    )
+                select
+                    s.id,
+                    s.raw_value,
+                    s.background,
+                    -- The append with null is silly but crucial to ensure that the
+                    -- cross join in `latest_cells_with_mention` returns all cells
+                    -- not just those that reference another cell
+                    ARRAY_APPEND(mentions(s.raw_value), null) as mentioned_cell_ids
+                from
+                    spreadsheet_data s
+                        join max_ts_per_cell mt on s.id = mt.id and s.ts = mt.max_ts;
+                
+                -- List all mentioned ids per latest cell
+                create view latest_cells_with_mentions as
+                select
+                    s.id,
+                    s.raw_value,
+                    s.background,
+                    m.mentioned_id
+                from
+                    latest_cells s, unnest(s.mentioned_cell_ids) as m(mentioned_id);
+                
+                -- Like latest_cells_with_mentions, but enrich it with values of mentioned cells
+                create view mentions_with_values as
+                select
+                    m.id,
+                    m.raw_value,
+                    m.background,
+                    m.mentioned_id,
+                    sc.computed_value as mentioned_value
+                from
+                    latest_cells_with_mentions m
+                left join
+                    spreadsheet_view sc on m.mentioned_id = sc.id;
+                
+                -- We aggregate mentioned values and ids back into arrays
+                create local view mentions_aggregated as
+                select
+                    id,
+                    raw_value,
+                    background,
+                    ARRAY_AGG(mentioned_id) as mentions_ids,
+                    ARRAY_AGG(mentioned_value) as mentions_values
+                from
+                    mentions_with_values
+                group by
+                    id,
+                    raw_value,
+                    background;
+                
+                -- Calculate the final spreadsheet by executing the UDF for the formula
+                create materialized view spreadsheet_view as
+                select
+                    id,
+                    raw_value,
+                    background,
+                    mentions_ids,
+                    mentions_values,
+                    cell_value(raw_value, mentions_ids, mentions_values) AS computed_value
+                from
+                    mentions_aggregated;""";
+        var ccs = this.getCCS(sql);
+        this.addRustTestCase(ccs);
+    }
+
     @Test
     public void testRecursive() {
         String sql = """

